codec/src/main/java/io/netty/handler/codec/DelimitedLengthFieldBasedFrameDecoder.java

```diff
@@ -0,0 +1,179 @@
+    private long frameLength;
+
+    public DelimitedLengthFieldBasedFrameDecoder(
+            int maxFrameLength,
+            int lengthAdjustment,
+            boolean failFast,
+            ByteBuf delimiter,
+            Charset lengthFieldCharset,
+            boolean trimLengthString
+    ) {
+        super(maxFrameLength, failFast);
+        if (delimiter == null) {
+            throw new NullPointerException("delimiter");
+        }
+        if (!delimiter.isReadable()) {
+            throw new IllegalArgumentException("empty delimiter");
+        }
+
+        this.delimiter = delimiter.slice(delimiter.readerIndex(), delimiter.readableBytes());
+        this.lengthFieldCharset = lengthFieldCharset;
+
+        this.lengthAdjustment = lengthAdjustment;
+        this.trimLengthString = trimLengthString;
+    }
+
+    /**
+     * Returns the number of bytes between the readerIndex of the haystack and
+     * the first needle found in the haystack.  -1 is returned if no needle is
+     * found in the haystack.
+     * <p>
+     * For now, this delegates to the implementation within
+     * {@link DelimiterBasedFrameDecoder#indexOf(ByteBuf, ByteBuf)}
+     *
+     * @param haystack the buffer to search for an occurence of the needle
+     * @param needle the buffer to find in the haystack
+     *
+     * @return the first byte index of the needle found in the haystack, or -1 if not found
+     */
+    protected int indexOf(ByteBuf haystack, ByteBuf needle) {
+        return DelimiterBasedFrameDecoder.indexOf(haystack, needle);
+    }
+
+    /**
+     * Create a frame out of the {@link ByteBuf} and return it.
+     *
+     * @param ctx the {@link ChannelHandlerContext} which this {@link ByteToMessageDecoder} belongs to
+     * @param in the {@link ByteBuf} from which to read data
+     *
+     * @return the {@link ByteBuf} which represents the frame or {@code null} if no frame could be created.
+     */
+    @Override
+    protected Object decode(ChannelHandlerContext ctx, ByteBuf in) throws Exception {
+        if (discardingTooLongFrame) {
+            long bytesToDiscard = this.bytesToDiscard;
+            int localBytesToDiscard = (int) Math.min(bytesToDiscard, in.readableBytes());
+            in.skipBytes(localBytesToDiscard);
+            bytesToDiscard -= localBytesToDiscard;
+            this.bytesToDiscard = bytesToDiscard;
+
+            failIfNecessary(false);
+
+            // if exception was not thrown, reader has been advanced by readable bytes already, so return null to
+            // continue processing too long frame
+            // TODO: check whether also needed in LengthFieldBasedFrameDecoder; I believe it is
+            return null;
+        }
+
+        if (consumingLength) {
+            int delimIndex = indexOf(in, delimiter);
+            if (delimIndex < 0) {
+                return null;
+            }
+
+            final String lengthStr = in.toString(in.readerIndex(), delimIndex, lengthFieldCharset);
```

Comment:
Second parameter in the `toString()` is a `length`:
`in.toString(in.readerIndex(), delimIndex - in.readerIndex(), lengthFieldCharset);`
It's not catched in the tests?

I still suggest to extract the reading and parsing of a frame length into separate method. And it makes sense to add the check for a big difference between `in.readerIndex()` and `delimIndex` (to prevent a possible DDoS):

```java
// Max size of the frame length field, even taking into a whitespaces presence
private static final int MAX_LENGTH_OF_LENGTH_FIELD = 64;

...

protected int readFrameLength(ByteBuf in, int from, int to) {
    final int lengthStrSize = from - to;
    if (lengthStrSize < 1 || lengthStrSize > MAX_LENGTH_OF_LENGTH_FIELD) {
        throw new CorruptedFrameException(String.format(
            "wrong size of the pre-adjustment length field: %d (expected between 1 and %d)",
            lengthStrSize, MAX_LENGTH_OF_LENGTH_FIELD));
    }

    final String lengthStr = in.toString(from, lengthStrSize, lengthFieldCharset);
    final int lengthValue;
    try {
        lengthValue = Integer.parseInt(trimLengthString? lengthStr.trim() : lengthStr);
    } catch (NumberFormatException nfe) {
        throw new CorruptedFrameException(
            String.format("invalid length field decoded (in %s charset): %s",
                lengthFieldCharset.name(), lengthStr), nfe);
    }
	
    if (lengthValue < 0) {
        throw new CorruptedFrameException("negative pre-adjustment length field: " + lengthStr);
    }
	
    return lengthValue;
}

...

frameLength = readFrameLength(in, in.readerIndex(), delimIndex);
```